keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -656,73 +661,287 @@ ways.  One is that we are dispatching requests to be served rather
-ideal.  In plain fair queuing the bound is one packet; in our version
-it is C requests.
-
-To support efficiently making the necessary adjustments once a
-request’s actual service time is known, the virtual finish time of a
-request and the last virtual finish time of a queue are not
-represented directly but instead computed from queue length, request
-position in the queue, and an alternate state variable that holds the
-queue’s virtual start time.  While the queue is empty and has no
-requests executing: the value of its virtual start time variable is
-ignored and its last virtual finish time is considered to be in the
-virtual past.  When a request arrives to an empty queue with no
-requests executing, the queue’s virtual start time is set to `now()`.
-The virtual finish time of request number J in the queue (counting
-from J=1 for the head) is J * G + (virtual start time).  While the
-queue is non-empty: the last virtual finish time of the queue is the
-virtual finish time of the last request in the queue.  While the queue
-is empty and has a request executing: the last virtual finish time is
-the queue’s virtual start time.  When a request is dequeued for
-service the queue’s virtual start time is advanced by G.  When a
-request finishes being served, and the actual service time was S, the
-queue’s virtual start time is decremented by G - S.
+done being served.
+
+To show how to cope with those differences, we start by re-casting the
+original fair queuing story in a way that is more amenable to the
+needed changes and has the same implementation.  Then we work through
+a series of three changes.
+
+##### The original story
+
+We have a collection of FIFO queues of packets to be transmitted, one
+packet at a time, over one link that goes at a fixed rate of
+`mu_single` (quantified in bits/second).  We imagine a virtual world
+called bit-by-bit round-robin (BR), in which the transmitter serves
+the queues in a round-robin fashion but advances from one queue to the
+next not on packet boundaries but rather on bit boundaries.  We define
+a function of time R(t) that counts the number of "rounds" completed,
+where a round is sending one bit from each non-empty queue.  Thus R(t)
+is
+
+```
+Integral[from tau=start_of_story to tau=t] (
+    if NAC(tau) > 0 then mu_single / NAC(tau) else 0) dtau
+```
+
+where `NAC(a time)` is the number of "active" or "non-empty" queues at
+that time, defined precisely below.
+
+We define start and finish R values for each packet:
+
+```
+S(i,j) = max(F(i,j-1), R(t_arrive(i,j)))
+F(i,j) = S(i,j) + len(i,j) 
+```
+
+where:
+- `S(i,j)` is the R value for the start of the j'th packet of queue i,
+- `F(i,j)` is the R value for the finish of the j'th packet of queue i,
+- `len(i,j)` is the number of bits in that packet, and
+- `t_arrive(i,j)` is the time when that packet arrived.
+
+We define `NAC(t)` to be the number of queues `i` for which `R(t) <=
+F(i,j)` where `j` is the last packet of queue `i` to arrive at or
```

Comment:
You are proposing a different definition of NAC than appears in the paper (focusing on the packet that is being transmitted right now in BR rather than the last arrival before t).  I think I can prove that the two definitions are equivalent, but I would rather not have to.

Reply:
I tried an intermediate rewording that I hope is clear enough and more obviously equivalent to the definition in the original paper.